### DISCUSSION TITLE
`encode` and `decode` functions should all have the same category

### DISCUSSION THREAD

**Discussion by kaathewisegit** [nushell/nushell#13419] [General]
Currently the `decode`/`encode` functions have a String category, `base64` versions use Hash, and the `hex` ones use `formats`.  I feel like all of those (or at least both `base64` and `hex`) should have the same category.

**Comment by fdncred** [comment]
We used to have all these commands together in one base name command until someone complained that hashing and encoding/decoding are separate and different things. This is why they're apart and categorized differently.

**Reply by kaathewisegit** [reply]
Huh, I see.  But I don't think Base64 encoding is hashing, is it?  Hash functions typically output a fixed-size identifier, which isn't usually reversible.  And `hash sha256` and `encode base64` seem like they are doing completely different things.  The former is creating a short "summary" of the data, while the latter encodes the data whole.

**Reply by kaathewisegit** [reply]
I do agree that Base64 & friends shouldn't be a part of `decode`, because `decode` is mostly concerned with text pages/encodings.

**Reply by fdncred** [reply]
ya, i'm not saying what we have is perfect. just citing some nushell history.

**Reply by kaathewisegit** [reply]
Got it!  The reason I raised this is because I'm making a decode/encode plugin with `base32`, `z32`, and such, and I was wondering which category I should use.  I decided to go with `formats` for all of them.  In Your opinion, would it be a good idea to open a PR to change `base64` and `hex` commands to the `formats` category too?

**Reply by fdncred** [reply]
all the basexx commands are encodings but formats is probably ok for now. it's more accurate than hash i guess.

**Reply by fdncred** [reply]
I wonder if it would be a better use of nushell to have a `base` command that takes a parameter for the base to encode things to?
https://github.com/multiformats/multibase
https://github.com/multiformats/rust-multibase

**Reply by kaathewisegit** [reply]
I like the idea and want to implement it.  But I see two problems:

- Each encoding has different switches.  Base32 has two: lowercase variant and no-padding variant.  Same goes for its hex version.  I feel that having 8 slightly different variants of Base32 would be confusing to type.

- Library situation is not great.  Multibase doesn't seem to be very maintained, and it's also missing Crockford.  If it was up to me, I'd stick to [`data-encoding`](https://lib.rs/crates/data-encoding), which is well-established, and only use hex/base32/base64 in Nushell.

**Reply by fdncred** [reply]
ya, i saw data-encoding after i posted the multibase but i figured you'd get the idea i was trying to make anyway, which you did. unfortunately, there seem to be many other "base" encodings that data-encoding doesn't support, if i'm reading it correctly. even so, using one library that adds a few may be a win?

**Reply by kaathewisegit** [reply]
Yeah, `data_encoding` focuses on the more mainstream ones.  I, honestly, wouldn't be comfortable adding others into the mainline Nushell, since there are a lot of nuances to get right (padding along is a hassle, the creator of `base-x` even put a disclaimer, advising against using Their library for Base32/Base64).  And, as I found out, some of them tend to panic on invalid input.  In that regard `base64` and `data-encoding` seem like two of the most established and polished libraries.

I'll draft a PR and we'll see how it'll look in practice
